TiDB panic in (*memoryUsageAlarm).recordSQL

## Bug Report

Please answer these questions before submitting your issue. Thanks!
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2736fa4]goroutine 2180 [running]:
github.com/pingcap/tidb/util/memory.(*Tracker).MaxConsumed(...)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/tracker.go:406
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).recordSQL.func3(0x0?, 0x14e1684?)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:214 +0x34
sort.order2_func(...)
        /usr/local/go/src/sort/zsortfunc.go:299
sort.median_func({0x4dacba6068?, 0x44205a2620?}, 0x3, 0x6, 0x9, 0x4dacba5c88)
        /usr/local/go/src/sort/zsortfunc.go:309 +0x88
sort.choosePivot_func({0x4dacba6068?, 0x44205a2620?}, 0x40?, 0x191?)
        /usr/local/go/src/sort/zsortfunc.go:284 +0xe8
sort.pdqsort_func({0x4dacba6068?, 0x44205a2620?}, 0x0?, 0x191?, 0x4dacba5db8?)
        /usr/local/go/src/sort/zsortfunc.go:89 +0xa4
sort.Slice({0x3453ba0, 0x460a896030}, 0x1e3?)
        /usr/local/go/src/sort/slice.go:23 +0x94
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).recordSQL.func2(0x454e4f4600?)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:187 +0x6c
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).recordSQL(0x4dacba7f20, {0x4018758?, 0x401072aee0?})
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:213 +0x4bc
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).doRecord(0x4dacba7f20, 0x1517230?, 0x6026780?, {0x4018758, 0x401072aee0})
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:145 +0x73c
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).alarm4ExcessiveMemUsage(0x4dacba7f20, {0x4018758, 0x401072aee0})
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:121 +0x390
github.com/pingcap/tidb/util/expensivequery.(*Handle).Run(0x40007168b8)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/expensivequery/expensivequery.go:91 +0x1b8
created by main.createServer
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/tidb-server/main.go:730 +0x3a8
```

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/util/expensivequery/expensivequerey_test.go
+++ b/util/expensivequery/expensivequerey_test.go
@@ -79,3 +79,26 @@ func TestLogFormat(t *testing.T) {
 	logFields = genLogFields(costTime, info)
 	assert.Equal(t, "select * from table where `a` > ?", logFields[6].String)
 }
+
+func TestGetCurrentMemoryUsage(t *testing.T) {
+	mem := memory.NewTracker(-1, -1)
+	memUsage := int64(1<<30 + 1<<29 + 1<<28 + 1<<27)
+	mem.Consume(memUsage)
+	var refCount stmtctx.ReferenceCount = 0
+	stmtCtx := &stmtctx.StatementContext{
+		MemTracker: mem,
+	}
+	// case 1, refCount is nil
+	assert.Equal(t, int64(0), getCurrentMemoryUsage(nil, stmtCtx))
+	// case 2, refCount is not nil
+	assert.Equal(t, memUsage, getCurrentMemoryUsage(&refCount, stmtCtx))
+	// case 3, refCount is not nil, but refCount.TryIncrease() return false
+	assert.Equal(t, true, refCount.TryFreeze())
+	assert.Equal(t, int64(0), getCurrentMemoryUsage(&refCount, stmtCtx))
+	refCount.UnFreeze()
+	// case 4, stmtCtx is reset
+	assert.Equal(t, true, refCount.TryFreeze())
+	*stmtCtx = stmtctx.StatementContext{}
+	refCount.UnFreeze()
+	assert.Equal(t, int64(0), getCurrentMemoryUsage(&refCount, stmtCtx))
+}

--- a/util/expensivequery/expensivequerey_test.go
+++ b/util/expensivequery/expensivequerey_test.go
@@ -18,11 +18,13 @@ import (
 	"testing"
 	"time"
 
+	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/testkit/testsetup"
 	"github.com/pingcap/tidb/util"
 	"github.com/pingcap/tidb/util/memory"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"go.uber.org/goleak"
 )
 
@@ -92,11 +94,17 @@ func TestGetCurrentMemoryUsage(t *testing.T) {
 	assert.Equal(t, int64(0), getCurrentMemoryUsage(nil, stmtCtx))
 	// case 2, refCount is not nil
 	assert.Equal(t, memUsage, getCurrentMemoryUsage(&refCount, stmtCtx))
-	// case 3, refCount is not nil, but refCount.TryIncrease() return false
+	// case 3, panic inside getCurrentMemoryUsage
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/util/expensivequery/panicWhenGetCurrentMemoryUsage", `return`))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/util/expensivequery/panicWhenGetCurrentMemoryUsage"))
+	}()
+	assert.Equal(t, int64(0), getCurrentMemoryUsage(&refCount, stmtCtx))
+	// case 4, refCount is not nil, but refCount.TryIncrease() return false
 	assert.Equal(t, true, refCount.TryFreeze())
 	assert.Equal(t, int64(0), getCurrentMemoryUsage(&refCount, stmtCtx))
 	refCount.UnFreeze()
-	// case 4, stmtCtx is reset
+	// case 5, stmtCtx is reset
 	assert.Equal(t, true, refCount.TryFreeze())
 	*stmtCtx = stmtctx.StatementContext{}
 	refCount.UnFreeze()

--- a/util/expensivequery/expensivequerey_test.go
+++ b/util/expensivequery/expensivequerey_test.go
@@ -94,19 +94,23 @@ func TestGetCurrentMemoryUsage(t *testing.T) {
 	assert.Equal(t, int64(0), getCurrentMemoryUsage(nil, stmtCtx))
 	// case 2, refCount is not nil
 	assert.Equal(t, memUsage, getCurrentMemoryUsage(&refCount, stmtCtx))
+	assert.Equal(t, stmtctx.ReferenceCount(0), refCount)
 	// case 3, panic inside getCurrentMemoryUsage
 	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/util/expensivequery/panicWhenGetCurrentMemoryUsage", `return`))
 	defer func() {
 		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/util/expensivequery/panicWhenGetCurrentMemoryUsage"))
 	}()
 	assert.Equal(t, int64(0), getCurrentMemoryUsage(&refCount, stmtCtx))
+	assert.Equal(t, stmtctx.ReferenceCount(0), refCount)
 	// case 4, refCount is not nil, but refCount.TryIncrease() return false
 	assert.Equal(t, true, refCount.TryFreeze())
 	assert.Equal(t, int64(0), getCurrentMemoryUsage(&refCount, stmtCtx))
 	refCount.UnFreeze()
+	assert.Equal(t, stmtctx.ReferenceCount(0), refCount)
 	// case 5, stmtCtx is reset
 	assert.Equal(t, true, refCount.TryFreeze())
 	*stmtCtx = stmtctx.StatementContext{}
 	refCount.UnFreeze()
 	assert.Equal(t, int64(0), getCurrentMemoryUsage(&refCount, stmtCtx))
+	assert.Equal(t, stmtctx.ReferenceCount(0), refCount)
 }

--- a/util/expensivequery/memory_usage_alarm.go
+++ b/util/expensivequery/memory_usage_alarm.go
@@ -175,9 +175,7 @@ func getCurrentMemoryUsage(refCount *stmtctx.ReferenceCount, stmtCtx *stmtctx.St
 		}()
 		if tracker := stmtCtx.MemTracker; tracker != nil {
 			ret = tracker.MaxConsumed()
-			if _, _err_ := failpoint.Eval(_curpkg_("panicWhenGetCurrentMemoryUsage")); _err_ == nil {
-				panic("panic when get current memory usage")
-			}
+			failpoint.Inject("panicWhenGetCurrentMemoryUsage", func() { panic("panic when get current memory usage") })
 		}
 	}
 	return

--- a/util/expensivequery/memory_usage_alarm.go
+++ b/util/expensivequery/memory_usage_alarm.go
@@ -231,9 +231,9 @@ func (record *memoryUsageAlarm) recordSQL(sm util.SessionManager) {
 
 	_, err = f.WriteString("The 10 SQLs with the most memory usage for OOM analysis\n")
 	printTop10(func(i, j int) bool {
-		i_consume := getCurrentMemoryUsage(pinfo[i].RefCountOfStmtCtx, pinfo[i].StmtCtx)
-		j_consume := getCurrentMemoryUsage(pinfo[j].RefCountOfStmtCtx, pinfo[j].StmtCtx)
-		return i_consume > j_consume
+		iConsume := getCurrentMemoryUsage(pinfo[i].RefCountOfStmtCtx, pinfo[i].StmtCtx)
+		jConsume := getCurrentMemoryUsage(pinfo[j].RefCountOfStmtCtx, pinfo[j].StmtCtx)
+		return iConsume > jConsume
 	})
 
 	_, err = f.WriteString("The 10 SQLs with the most time usage for OOM analysis\n")

--- a/util/expensivequery/memory_usage_alarm.go
+++ b/util/expensivequery/memory_usage_alarm.go
@@ -175,7 +175,9 @@ func getCurrentMemoryUsage(refCount *stmtctx.ReferenceCount, stmtCtx *stmtctx.St
 		}()
 		if tracker := stmtCtx.MemTracker; tracker != nil {
 			ret = tracker.MaxConsumed()
-			failpoint.Inject("panicWhenGetCurrentMemoryUsage", func() { panic("panic when get current memory usage") })
+			if _, _err_ := failpoint.Eval(_curpkg_("panicWhenGetCurrentMemoryUsage")); _err_ == nil {
+				panic("panic when get current memory usage")
+			}
 		}
 	}
 	return

--- a/util/expensivequery/memory_usage_alarm.go
+++ b/util/expensivequery/memory_usage_alarm.go
@@ -25,6 +25,7 @@ import (
 	"time"
 
 	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/util"
 	"github.com/pingcap/tidb/util/disk"
 	"github.com/pingcap/tidb/util/logutil"
@@ -161,6 +162,17 @@ func (record *memoryUsageAlarm) doRecord(memUsage uint64, instanceMemoryUsage ui
 	}
 }
 
+func getCurrentMemoryUsage(refcount *stmtctx.ReferenceCount, stmtCtx *stmtctx.StatementContext) int64 {
+	if refcount != nil && refcount.TryIncrease() {
+		defer refcount.Decrease()
+		// note there is still data race when reading MemTracker, but this is the best effort without too many changes
+		if tracker := stmtCtx.MemTracker; tracker != nil {
+			return tracker.MaxConsumed()
+		}
+	}
+	return 0
+}
+
 func (record *memoryUsageAlarm) recordSQL(sm util.SessionManager) {
 	processInfo := sm.ShowProcessList()
 	pinfo := make([]*util.ProcessInfo, 0, len(processInfo))
@@ -211,7 +223,9 @@ func (record *memoryUsageAlarm) recordSQL(sm util.SessionManager) {
 
 	_, err = f.WriteString("The 10 SQLs with the most memory usage for OOM analysis\n")
 	printTop10(func(i, j int) bool {
-		return pinfo[i].StmtCtx.MemTracker.MaxConsumed() > pinfo[j].StmtCtx.MemTracker.MaxConsumed()
+		i_consume := getCurrentMemoryUsage(pinfo[i].RefCountOfStmtCtx, pinfo[i].StmtCtx)
+		j_consume := getCurrentMemoryUsage(pinfo[j].RefCountOfStmtCtx, pinfo[j].StmtCtx)
+		return i_consume > j_consume
 	})
 
 	_, err = f.WriteString("The 10 SQLs with the most time usage for OOM analysis\n")

--- a/util/expensivequery/memory_usage_alarm.go
+++ b/util/expensivequery/memory_usage_alarm.go
@@ -24,6 +24,7 @@ import (
 	"strings"
 	"time"
 
+	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/config"
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/util"
@@ -162,15 +163,22 @@ func (record *memoryUsageAlarm) doRecord(memUsage uint64, instanceMemoryUsage ui
 	}
 }
 
-func getCurrentMemoryUsage(refcount *stmtctx.ReferenceCount, stmtCtx *stmtctx.StatementContext) int64 {
-	if refcount != nil && refcount.TryIncrease() {
-		defer refcount.Decrease()
-		// note there is still data race when reading MemTracker, but this is the best effort without too many changes
+func getCurrentMemoryUsage(refCount *stmtctx.ReferenceCount, stmtCtx *stmtctx.StatementContext) (ret int64) {
+	ret = 0
+	if refCount != nil && refCount.TryIncrease() {
+		defer func() {
+			// even if we increase the refCount, reading MemTracker is still not thread-safe, so we add recover here
+			if r := recover(); r != nil {
+				ret = 0
+			}
+			refCount.Decrease()
+		}()
 		if tracker := stmtCtx.MemTracker; tracker != nil {
-			return tracker.MaxConsumed()
+			failpoint.Inject("panicWhenGetCurrentMemoryUsage", func() { panic("panic when get current memory usage") })
+			ret = tracker.MaxConsumed()
 		}
 	}
-	return 0
+	return
 }
 
 func (record *memoryUsageAlarm) recordSQL(sm util.SessionManager) {

--- a/util/expensivequery/memory_usage_alarm.go
+++ b/util/expensivequery/memory_usage_alarm.go
@@ -174,8 +174,8 @@ func getCurrentMemoryUsage(refCount *stmtctx.ReferenceCount, stmtCtx *stmtctx.St
 			refCount.Decrease()
 		}()
 		if tracker := stmtCtx.MemTracker; tracker != nil {
-			failpoint.Inject("panicWhenGetCurrentMemoryUsage", func() { panic("panic when get current memory usage") })
 			ret = tracker.MaxConsumed()
+			failpoint.Inject("panicWhenGetCurrentMemoryUsage", func() { panic("panic when get current memory usage") })
 		}
 	}
 	return

